### PR TITLE
test(flatten_cfg): unit test and end to end test for #1792

### PR DESCRIPTION
**Pull request by ludamad**
# Description

## Problem\*

Resolves 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs [hunk 1]
```diff
@@ -1196,4 +1198,182 @@ mod test {
...
+        //     v43 = eq v42, Field 0
+        //     v44 = not v43
+        //     constrain v44
+        //     v46 = eq v15, u1 1
+        //     constrain v46
+        //     v48 = eq v20, u1 0
+        //     constrain v48
+        //     return
+        // }
+        let test_set_id = Id::test_new(7);
```
Comment (ludamad) [inline crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:1264]: GPT4 did the conversion, plus a bit of hand-massaging

Comment (ludamad) [reply crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs]: Simplified (by GPT4 of course I'm not a sucker)

**Comment by ludamad** [comment]
@kevaundray @jfecher figured out that the simplify code thinks 
```
        // fn main f1 {
        //     ... simplified ...
        //     v_false = u1 0
        //     jmpif v_false then: b_useless_branch, else: b_fallthrough
        //   b_useless_branch():
        //     jmp b2()
        //   b_fallthrough():
        //     constrain v_false // was incorrectly removed
        //     return
        // }
```

the constrain here can be removed on flatten_cfg, but I didn't debug further.

**Comment by jfecher** [comment]
Can we open this as an issue in the future? Unless the PR already has the fix for the bug included I think it is better to file the bug first.

**Comment by ludamad** [comment]
Sure, although surely this needs the fix any way and then it has an issue? This can't be merged without it

**Comment by jfecher** [comment]
> Sure, although surely this needs the fix any way and then it has an issue? This can't be merged without it. I thought of this more as a hand off PR

Sorry, I misinterpreted this PR at first. I didn't see the existing linked issue #1792 somehow. I still think for cases like this we can include any testing code in the issue and just make sure to add it in the same PR the issue is fixed in. That way we have assurance the PR fixes the issue in question compared to adding the test afterward.

**Comment by ludamad** [comment]
Really I should have probably slapped DO NOT MERGE on this, it was meant to be used as you did, cherry-picked into the fix PR. And it is in the issue by virtue of mentioning the issue

**Pull request closed without merging**